### ISSUE TITLE
Add a flag to treat warnings as errors

### ISSUE THREAD

**Issue by notorand-it** [opentofu/opentofu#757]
### OpenTofu Version

```shell
1.6.0-alpha2
```

### Use Cases

In a number of environments (like mine), deprecation warnings are to be fixed before commit.

### Attempted Solutions

Only those ones that involve adding complex `grep` regexes and shell scripting.

### Proposal

Add a flag like `-warnings-are-errors` or `-pedantic` to let the tool fail with non-zero return value also in case of warnings.
Default behavior would not treat warnings as error, thus retaining backward compatibility.

**Comment by hainenber** [comment]
Hi there, I'd love to work on this. Can I be assigned? Thanks!

**Comment by notorand-it** [comment]
> Thanks for creating the issue @notorand-it!
> 
> I personally like this proposal very much. I think in C/C++ compilers this is usually `-werr`?

I don't really mind how the flag will be called, as long as it does its work... ;-)

**Comment by janosdebugs** [comment]
Hey folks, thank you for your input and apologies for the late reply.. We have discussed this issue in the core team and would be happy about a PR for this issue with a slight preference for the `-pedantic` option. @hainenber would you still like to work on this?

**Comment by hainenber** [comment]
Yes, please help assign to me if possible!

**Comment by Yantrio** [comment]
I've assigned you to this issue. Thank you @hainenber

**Comment by siddharthasonker95** [comment]
Hey @hainenber, please let us know if you are working on this issue, or should we reassign it?

**Comment by hainenber** [comment]
Yes, you can reassign to other folks

**Comment by bunniseng** [comment]
Hi @janosdebugs 

Please assign this issue to me.

Thanks

**Comment by bunniseng** [comment]
Hey @cam72cam and @janosdebugs

I have a question about the implementation, do we want this flag to be a global flag or for specific commands like plan and apply ?

Thanks

**Comment by notorand-it** [comment]
I would implement this whenever you can have warnings. Validating, planning and applying firstly.

I am not sure whether formatting and destroying can also generate warning, though.

**Comment by bunniseng** [comment]
I'm a little bit time poor over the next week or so with other commitments, I'll continue on with this shortly.

**Comment by bunniseng** [comment]
Hi @janosdebugs  and @notorand-it 

Just looking for a little bit more clarity on the behaviour of this feature enhancement before I get too stuck in.

Are we saying that when we have set the `-pedantic` global option and a warning is produced, we are expecting the application to immediately exit with a non-zero return code ? or are we saying to treat warnings as errors and to change the output colour to red and write the warning to stderr as well as return the non-zero return code?

Thanks

**Comment by cam72cam** [comment]
I believe the second option, "change the output colour to red (in the case of a colourised message) and write the warning to stderr as well as return the non-zero return code?"

**Comment by RLRabinowitz** [comment]
I'd imagine we might need a mechanism here to be able to "ignore" specific warnings. That would allow users to ignore specific unwanted warnings that cannot really be addressed, and prevent the action from failing

**Comment by bunniseng** [comment]
Yeah so that can be done by mapping codes to warnings and having an additional flag to indicate warnings to ignore such as `--ignore-warnings=WARN100,WARN200` etc.

@cam72cam can you provide any insight on what you think about this and whether it should be included with this piece of work.

**Comment by cam72cam** [comment]
I think we should implement this as originally described and create a followup issue to track include/exclude of warnings in pedantic mode.

**Comment by bunniseng** [comment]
PR [1726](https://github.com/opentofu/opentofu/pull/1726) is ready for review.

Screenshot shows running a command which produces a warning and an exit code of 0
![normal output](https://github.com/user-attachments/assets/1978c66b-26b0-416d-b917-830ed9eed43d)

Screenshot demonstrates running a command with the pedantic global option toggled which produces an error and an exit code of 1.
![pedantic output](https://github.com/user-attachments/assets/2a76e043-d6d5-491b-ab6f-1f16774c5ff2)

I have covered legacy UI component, new view component and json output.

**Comment by cam72cam** [comment]
Excellent, I will review this in detail this week!

**Comment by notorand-it** [comment]
Hmm... You promoted a warning to error instead of treating the warning as an error. Which is not the same IMHO.

**Comment by cam72cam** [comment]
The question for @opentofu/core-engineers is which approach we take.

To quote the issue as accepted:
>  Default behavior would not treat warnings as error, thus retaining backward compatibility.

There's also discussion of adopting what C/C++ does with `-Werror`, explicitly turning any warnings into errors. This approach is what @bunniseng implemented.

I'll try to get a discussion going next Monday with the core team.

**Comment by apparentlymart** [comment]
(The following is just my personal opinion and not an official decision from the OpenTofu Core team. Please don't immediately prepare or update any pull requests based on these ideas.)

When I consider the tradeoff above (about how to present the "promoted" warnings), the main concern that comes to my mind is this:

The person who is receiving the messages may not be the person who added the extra option to the command line in the automation environment, and they may not be someone who is already highly experienced with OpenTofu and thus able to infer from memory which of the messages are "real" errors and which of them are just warnings pretending to be errors. While sometimes that distinction won't matter, I do tend to want to prioritize having a reader of a message understand its cause and action as well as possible before consulting any other materials.

With that tradeoff in mind, I would propose the following compromise:

- Retain the on-screen marking of the individual diagnostics as warnings.
- If this option is set, then whenever preparing a final set of diagnostics for use by the CLI layer (`command` package, `backend` package, etc), add an additional Error diagnostics "just in time" which is explicit that warnings are being treated as errors and mentions why that is so.

An example of the potential additional message:

```
│ Error: Treating all warnings as errors
│
│ The command line option -pedantic forces treatment of the included warnings as errors.
```

This gives the reader direct feedback about what has happened any why it has happened, so they can decide how to proceed. Leaving the original warning messages still marked as warnings makes it clearer which of the diagnostic messages this additional error relates to, in case there are also some truly-error messages in the set for which removing `-pedantic` would not have any effect.

In particular, I suspect that with the initial form of this feature (without any mechanism to specify only a subset of warnings to "promote") there _will_ be situations where it's impossible to proceed without temporarily disabling this option to allow the correction of something that's causing a warning. Some warnings are caused directly or indirectly by something that's present in the prior state rather than in the configuration or on the command line, in which case successfully performing a plan and apply round (or some other workflow phase) would likely be the only way to clear the error.

Making it clear that this is something a team chose to impose on themselves and that they can choose to (temporarily or permanently) un-impose it to solve a problem feels important to me, so that a person in that situation doesn't waste too much time stuck in that state before they get a hint on how to escape from it.

**Comment by ucfchandra** [comment]
@bunniseng is this issue still active? Would you like help?

**Comment by nitrocode** [comment]
Related to

- https://github.com/hashicorp/terraform/issues/32289

**Comment by apparentlymart** [comment]
It seems like this issue got a bit "stuck" because OpenTofu's existing use of warning diagnostics is too broad for making a general rule that _all_ warnings are to be treated as errors. For example, that would mean that `-target` and `-exclude` would always be forbidden in conjunction with this option because they _always_ generate warnings when used.

However, I notice that the original framing for this issue was about deprecation warnings in particular. I wonder then if we could compromise and say that the requirement is to treat _only_ deprecation warnings as errors, while leaving all other warnings as warnings for now (though we could potentially add support for other warning categories later if we learn of good reasons for them).

In principle we _can_ annotate diagnostics with additional metadata that we can use to trigger special behavior, and we already do that in a few cases. However, a specific challenge for this case is that provider plugins can potentially generate deprecation warnings but the provider protocol doesn't allow the provider to distinguish between deprecation-related and non-deprecation-related warnings. So if we took this compromise I expect we'd need to consider all _provider-generated_ warnings as if they are deprecation warnings. Although that's not universally true, I suspect it's still a better compromise than treating _all warnings whatsoever_ as if they are deprecation warnings, since there are far fewer situations where it's helpful for a provider to generate a warning.

Our new v1.10 features for marking input variables and output values as deprecated (from https://github.com/opentofu/opentofu/issues/1005) are handled by OpenTofu core, so it would be considerably easier to handle those as a special case distinguished from other warnings that OpenTofu itself generates.

---

For those who were interested in this issue, would you find it an acceptable compromise for the new option to treat only the following subset of warnings as errors?

- Any warning returned by a provider through the provider plugin protocol, regardless of whether it's talking about a deprecation situation.
- Any warning where OpenTofu is reporting that an input variable is deprecated.
- Any warning where OpenTofu is reporting that an output value is deprecated.
- Any other warnings we might introduce in future about OpenTofu's own features becoming deprecated. (I don't think there are currently any in this category, but that could change if new features lead to older features becoming deprecated.)

All other warnings would, for now, remain treated exclusively as warnings.

However, we'd design this in such a way that it allows for other subsets of warnings to potentially get similar treatment in future, such as an option syntax that takes a comma-separated list of category keywords where the first implementation would support only one category keyword "deprecation". (But let's focus on whether this compromise is even viable first, before we worry too much about the exact syntax details.)
